### PR TITLE
EVG-17882: Latency is incorrect with multiple ops reported in single …

### PR DESCRIPTION
**Pull request by mridul-augustine**
…event

PLEASE READ BEFORE DELETING

You MUST bump the package version in pyproject.toml before requesting a review on this PR. Merging this PR will push the package to pypi. We use semver for package versioning.

Read more about semver here: https://semver.org/

### CODE REVIEW

File: src/ftdc_tools/rollups/client_perf.py [hunk 1]
```diff
@@ -22,6 +22,14 @@ class Statistic:
...
 
 
+@dataclass
+class SummaryStatistic:
+    """A summary statistic."""
+
+    min: float
+    max: float
+
+
```
Comment (aggrand) [inline src/ftdc_tools/rollups/client_perf.py]: Aren't all the statistics we're calculating summary statistics?

File: src/ftdc_tools/rollups/client_perf.py [hunk 2]
```diff
@@ -58,13 +78,21 @@ def add_doc(self, doc: FTDCDoc) -> None:
         else:
             duration = float(doc["timers"]["duration"])
         extracted_duration = duration - self.previous_duration
-
+        number_of_ops = doc["counters"]["ops"] - self.previous_ops
+        number_of_ops = (
+            1 if number_of_ops == 0 else number_of_ops
+        )  # For multi-operation events that have same ops
+        self._operations_total = self._operations_total + number_of_ops
```
Comment (aggrand) [inline src/ftdc_tools/rollups/client_perf.py]: Do we need to add to this sum every time? The ops are cumulative so the line deleted further down of `self._operations_total = self.last_doc["counters"]["ops"]` should still work even for multi-operation events, so we can still use that at the end and not have `n` executions of this line

File: src/ftdc_tools/rollups/client_perf.py [hunk 3]
```diff
@@ -58,13 +78,21 @@ def add_doc(self, doc: FTDCDoc) -> None:
         else:
             duration = float(doc["timers"]["duration"])
         extracted_duration = duration - self.previous_duration
-
+        number_of_ops = doc["counters"]["ops"] - self.previous_ops
+        number_of_ops = (
+            1 if number_of_ops == 0 else number_of_ops
```
Comment (aggrand) [inline src/ftdc_tools/rollups/client_perf.py]: Why are we making number of ops 1 in this case? Shouldn't we keep 0 ops if 0 were reported?

File: src/ftdc_tools/rollups/client_perf.py [hunk 4]
```diff
@@ -82,7 +110,9 @@ def add_doc(self, doc: FTDCDoc) -> None:
         self._gauges_workers_max = min(
             self._gauges_workers_max, doc["gauges"]["workers"]
         )
-        self._extracted_durations.append(extracted_duration)
+        self._extracted_durations = self._extracted_durations + number_of_ops * [
+            extracted_duration / number_of_ops
+        ]
```
Comment (aggrand) [inline src/ftdc_tools/rollups/client_perf.py]: While I like the elegance here, I think this list concatenation might be less efficient (both memory and CPU-wise) than generating the new list and repeatedly appending to `self._extracted_durations`. It'd be similar to when we were using dataframes: a new one being allocated and repeatedly iterating over it which killed performance

**Review by aggrand — CHANGES_REQUESTED** [review]
This is looking great just have a few thoughts/questions

File: src/ftdc_tools/rollups/client_perf.py (hunk 2 above)
Comment (mridul-augustine) [reply src/ftdc_tools/rollups/client_perf.py]: This is to handle cases like below where all ops are run in parallel and they are all 0.

image

File: src/ftdc_tools/rollups/client_perf.py (hunk 3 above)
Comment (mridul-augustine) [reply src/ftdc_tools/rollups/client_perf.py]: To handle the same case mentioned in the previous comment.

File: src/ftdc_tools/rollups/client_perf.py (hunk 2 above)
Comment (aggrand) [reply src/ftdc_tools/rollups/client_perf.py]: @mridul-augustine I believed that's a bugged FTDC file from before [TIG-2804](https://jira.mongodb.org/browse/TIG-2804); I think all current FTDC files should be incrementing `counters.ops` when there is an operation, and if that's not happening then there is misreporting going on somewhere that should be addressed

File: src/ftdc_tools/rollups/client_perf.py (hunk 1 above)
Comment (mridul-augustine) [reply src/ftdc_tools/rollups/client_perf.py]: It's Summary of summary Stats :D.

Removed it!

File: src/ftdc_tools/rollups/client_perf.py (hunk 2 above)
Comment (mridul-augustine) [reply src/ftdc_tools/rollups/client_perf.py]: Got it. Will fix it

File: src/ftdc_tools/rollups/client_perf.py (hunk 3 above)
Comment (mridul-augustine) [reply src/ftdc_tools/rollups/client_perf.py]: fixing this as well

File: src/ftdc_tools/rollups/client_perf.py [hunk 5]
```diff
@@ -58,13 +64,20 @@ def add_doc(self, doc: FTDCDoc) -> None:
...
+            op_duration = extracted_duration
+            if number_of_ops > 1:
+                op_duration = extracted_duration / number_of_ops
+            self._max_latency = self._min_latency = op_duration
         else:
-            self._min_duration = min(self._min_duration, extracted_duration)
-            self._max_duration = max(self._max_duration, extracted_duration)
+            if number_of_ops == 0:
+                raise ValueError(
+                    "Invalid FTDC data. Number of ops not icremented. There is a possible misreporting for perf data."
```
Comment (aggrand) [inline src/ftdc_tools/rollups/client_perf.py:77]: (nit) incremented

**Review by aggrand — APPROVED** [review]

**Pull request merged**